ui/app/js/controllers/required_signoffs_modal_base_controller.js

```diff
@@ -0,0 +1,349 @@
+      service = ProductRequiredSignoffs;
+    }
+    else if ($scope.mode === "permissions") {
+      service = PermissionsRequiredSignoffs;
+    }
+    else {
+      $scope.errors["exception"] = "Couldn't detect mode";
+      return;
+    }
+
+    $scope.saving = true;
+    CSRF.getToken()
+    .then(function(csrf_token) {
+      var deferreds = {};
+      var promises = [];
+      var first = true;
+
+      var current_role_names = [];
+      var new_role_names = [];
+      var all_role_names = [];
+      current_roles.forEach(function(rs) {
+        if (rs["role"] !== "") {
+          current_role_names.push(rs["role"]);
+          all_role_names.push(rs["role"]);
+        }
+      });
+      $scope.new_roles.forEach(function(rs) {
+        if (rs["role"] !== "") {
+          if (new_role_names.indexOf(rs["role"]) !== -1) {
+            $scope.errors["exception"] = "Multiple entries found for " + rs["role"] + ". Cannot continue.";
+            return;
+          }
+          new_role_names.push(rs["role"]);
+          if (all_role_names.indexOf(rs["role"]) === -1) {
+            all_role_names.push(rs["role"]);
+          }
+        }
+      });
+
+      all_role_names.forEach(function(role_name) {
+        // The safe thing to do is to assume signoff is required.
+        // This will get overridden below in the one case where we shouldn't.
+        var requires_signoff = true;
+        var action = null;
+        var pending = false;
+        var role = null;
+        var i;
+
+        // If the role is only in the new Roles, we'll need to create it.
+        if (current_role_names.indexOf(role_name) === -1 && new_role_names.indexOf(role_name) !== -1) {
+          action = "insert";
+          // If we're creating a new role, and there wasn't any already, signoff won't be required
+          // Also, if first is false, some earlier iteration of the loop already directly created
+          // a Signoff, so signoff will be required.
+          if (current_role_names.length === 0 && first) {
+            requires_signoff = false;
+          }
+          for (i = 0; i < $scope.new_roles.length; ++i) {
+            if ($scope.new_roles[i]["role"] === role_name) {
+              role = $scope.new_roles[i];
+              break;
+            }
+          }
+        }
+        // If the role is only in the current Roles, we'll be deleting it
+        else if (current_role_names.indexOf(role_name) !== -1 && new_role_names.indexOf(role_name) === -1) {
+          action = "delete";
+          for (i = 0; i < current_roles.length; ++i) {
+            if (current_roles[i]["role"] === role_name) {
+              pending = current_roles[i]["sc"] ? true : false;
+              role = current_roles[i];
+              break;
+            }
+          }
+        }
+        // Otherwise (it's in both), it's an update.
+        else {
+          action = "update";
+          // There must be a better way to do this, but I can't find it.
```

Comment:
"role" ends up getting used around https://github.com/mozilla/balrog/pull/253/files/5aa69637fd5becb2316f4cb86104fb25d4c9ae2f#diff-e729d8dd6ca1733c654cfbef188401beR264 when we start making http requests. Depending on what we're doing, we may need to pull signoffs_required or scheduled change information from it. Maybe it would be clearer to assign those to separate variables...